File: backend/src/projects/services/TestResultService.cs
using System;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using Testoria.Projects.DTO;
using Testoria.Projects.Exceptions;
using Testoria.Projects.Interfaces;

namespace Testoria.Projects.Services
{
    /// <summary>
    /// Сервис для управления результатами тестов
    /// </summary>
    public sealed class TestResultService
    {
        private readonly PrismaService _prismaService;
        private readonly IProjectService _projectService;

        /// <summary>
        /// Инициализирует новый экземпляр класса TestResultService
        /// </summary>
        /// <param name="prismaService">Сервис для работы с базой данных</param>
        /// <param name="projectService">Сервис для работы с проектами</param>
        public TestResultService(PrismaService prismaService, IProjectService projectService)
        {
            _prismaService = prismaService;
            _projectService = projectService;
        }

        /// <summary>
        /// Создает результат теста
        /// </summary>
        /// <param name="createTestResultDto">Данные для создания результата теста</param>
        /// <returns>Созданный результат теста</returns>
        /// <exception cref="ProjectNotFoundException">Если проект не найден</exception>
        public async Task<TestResultDto> CreateTestResultAsync(CreateTestResultDto createTestResultDto)
        {
            // Проверяем существование проекта
            var project = await _projectService.GetProjectByIdAsync(createTestResultDto.ProjectId);
            if (project == null)
            {
                throw new ProjectNotFoundException(createTestResultDto.ProjectId);
            }

            var testResult = await _prismaService.TestResult.CreateAsync(new()
            {
                ProjectId = createTestResultDto.ProjectId.ToString(),
                TestName = createTestResultDto.TestName,
                Status = createTestResultDto.Status,
                Duration = createTestResultDto.Duration,
                ErrorMessage = createTestResultDto.ErrorMessage,
                StackTrace = createTestResultDto.StackTrace,
                RawData = createTestResultDto.RawData
            });

            return new TestResultDto
            {
                Id = Guid.Parse(testResult.Id),
                ProjectId = Guid.Parse(testResult.ProjectId),
                TestName = testResult.TestName,
                Status = testResult.Status,
                Duration = testResult.Duration,
                ErrorMessage = testResult.ErrorMessage,
                StackTrace = testResult.StackTrace,
                RawData = testResult.RawData,
                CreatedAt = testResult.CreatedAt
            };
        }

        /// <summary>
        /// Создает результаты тестов из Jest JSON
        /// </summary>
        /// <param name="projectId">Идентификатор проекта</param>
        /// <param name="jestJson">JSON данные от Jest</param>
        /// <returns>Массив созданных результатов тестов</returns>
        public async Task<TestResultDto[]> CreateTestResultsFromJestAsync(Guid projectId, JsonElement jestJson)
        {
            var results = new List<TestResultDto>();

            // Проверяем существование проекта
            var project = await _projectService.GetProjectByIdAsync(projectId);
            if (project == null)
            {
                throw new ProjectNotFoundException(projectId);
            }

            // Парсим Jest JSON
            if (jestJson.TryGetProperty("testResults", out var testResults))
            {
                foreach (var testResult in testResults.EnumerateArray())
                {
                    if (testResult.TryGetProperty("assertionResults", out var assertionResults))
                    {
                        foreach (var assertion in assertionResults.EnumerateArray())
                        {
                            var createDto = new CreateTestResultDto
                            {
                                ProjectId = projectId,
                                TestName = assertion.TryGetProperty("title", out var title) ? title.GetString() ?? "Unknown Test" : "Unknown Test",
                                Status = GetStatusFromJest(assertion),
                                Duration = GetDurationFromJest(assertion),
                                ErrorMessage = GetErrorMessageFromJest(assertion),
                                StackTrace = GetStackTraceFromJest(assertion),
                                RawData = assertion
                            };

                            var result = await CreateTestResultAsync(createDto);
                            results.Add(result);
                        }
                    }
                }
            }

            return results.ToArray();
        }

        private static string GetStatusFromJest(JsonElement assertion)
        {
            if (assertion.TryGetProperty("status", out var status))
            {
                return status.GetString() switch
                {
                    "passed" => "passed",
                    "failed" => "failed",
                    "pending" => "skipped",
                    "todo" => "skipped",
                    _ => "unknown"
                };
            }
            return "unknown";
        }

        private static float? GetDurationFromJest(JsonElement assertion)
        {
            if (assertion.TryGetProperty("duration", out var duration))
            {
                return duration.GetSingle();
            }
            return null;
        }

        private static string? GetErrorMessageFromJest(JsonElement assertion)
        {
            if (assertion.TryGetProperty("failureMessages", out var failureMessages) && 
                failureMessages.GetArrayLength() > 0)
            {
                return failureMessages[0].GetString();
            }
            return null;
        }

        private static string? GetStackTraceFromJest(JsonElement assertion)
        {
            if (assertion.TryGetProperty("failureMessages", out var failureMessages) && 
                failureMessages.GetArrayLength() > 1)
            {
                return string.Join("\n", failureMessages.EnumerateArray().Skip(1).Select(f => f.GetString()));
            }
            return null;
        }
    }
}

File: backend/src/projects/services/test-result.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../../prisma/prisma.service';
import { IProjectService } from '../interfaces/project.service.interface';
import { CreateTestResultDto, TestResultDto, JestTestResultsDto } from '../dto/test-result.dto';
import { ProjectNotFoundException } from '../exceptions/project-not-found.exception';

/**
 * Сервис для управления результатами тестов
 */
@Injectable()
export class TestResultService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly projectService: IProjectService,
  ) {}

  /**
   * Создает результат теста
   * @param createTestResultDto Данные для создания результата теста
   * @returns Созданный результат теста
   * @throws ProjectNotFoundException Если проект не найден
   */
  async createTestResult(createTestResultDto: CreateTestResultDto): Promise<TestResultDto> {
    // Проверяем существование проекта
    const project = await this.projectService.getProjectById(createTestResultDto.projectId);
    if (!project) {
      throw new ProjectNotFoundException(createTestResultDto.projectId);
    }

    const testResult = await this.prisma.testResult.create({
      data: {
        projectId: createTestResultDto.projectId,
        testName: createTestResultDto.testName,
        status: createTestResultDto.status,
        duration: createTestResultDto.duration,
        errorMessage: createTestResultDto.errorMessage,
        stackTrace: createTestResultDto.stackTrace,
        rawData: createTestResultDto.rawData,
      },
    });

    return {
      id: testResult.id,
      projectId: testResult.projectId,
      testName: testResult.testName,
      status: testResult.status,
      duration: testResult.duration,
      errorMessage: testResult.errorMessage,
      stackTrace: testResult.stackTrace,
      rawData: testResult.rawData,
      createdAt: testResult.createdAt,
    };
  }

  /**
   * Создает результаты тестов из Jest JSON
   * @param projectId Идентификатор проекта
   * @param jestData JSON данные от Jest
   * @returns Массив созданных результатов тестов
   * @throws ProjectNotFoundException Если проект не найден
   */
  async createTestResultsFromJest(projectId: string, jestData: any): Promise<TestResultDto[]> {
    // Проверяем существование проекта
    const project = await this.projectService.getProjectById(projectId);
    if (!project) {
      throw new ProjectNotFoundException(projectId);
    }

    const results: TestResultDto[] = [];

    // Парсим Jest JSON
    if (jestData.testResults && Array.isArray(jestData.testResults)) {
      for (const testResult of jestData.testResults) {
        if (testResult.assertionResults && Array.isArray(testResult.assertionResults)) {
          for (const assertion of testResult.assertionResults) {
            const createDto: CreateTestResultDto = {
              projectId,
              testName: assertion.title || 'Unknown Test',
              status: this.getStatusFromJest(assertion),
              duration: assertion.duration,
              errorMessage: this.getErrorMessageFromJest(assertion),
              stackTrace: this.getStackTraceFromJest(assertion),
              rawData: assertion,
            };

            const result = await this.createTestResult(createDto);
            results.push(result);
          }
        }
      }
    }

    return results;
  }

  /**
   * Получает результаты тестов для проекта
   * @param projectId Идентификатор проекта
   * @returns Массив результатов тестов
   */
  async getTestResultsByProject(projectId: string): Promise<TestResultDto[]> {
    const testResults = await this.prisma.testResult.findMany({
      where: { projectId },
      orderBy: { createdAt: 'desc' },
    });

    return testResults.map(testResult => ({
      id: testResult.id,
      projectId: testResult.projectId,
      testName: testResult.testName,
      status: testResult.status,
      duration: testResult.duration,
      errorMessage: testResult.errorMessage,
      stackTrace: testResult.stackTrace,
      rawData: testResult.rawData,
      createdAt: testResult.createdAt,
    }));
  }

  /**
   * Получает результат теста по идентификатору
   * @param id Идентификатор результата теста
   * @returns Результат теста или null, если не найден
   */
  async getTestResultById(id: string): Promise<TestResultDto | null> {
    const testResult = await this.prisma.testResult.findUnique({
      where: { id },
    });

    if (!testResult) {
      return null;
    }

    return {
      id: testResult.id,
      projectId: testResult.projectId,
      testName: testResult.testName,
      status: testResult.status,
      duration: testResult.duration,
      errorMessage: testResult.errorMessage,
      stackTrace: testResult.stackTrace,
      rawData: testResult.rawData,
      createdAt: testResult.createdAt,
    };
  }

  /**
   * Преобразует статус Jest в внутренний формат
   * @param assertion Результат теста от Jest
   * @returns Статус теста
   */
  private getStatusFromJest(assertion: any): string {
    switch (assertion.status) {
      case 'passed':
        return 'passed';
      case 'failed':
        return 'failed';
      case 'pending':
      case 'todo':
        return 'skipped';
      default:
        return 'unknown';
    }
  }

  /**
   * Извлекает сообщение об ошибке из результата Jest
   * @param assertion Результат теста от Jest
   * @returns Сообщение об ошибке или null
   */
  private getErrorMessageFromJest(assertion: any): string | null {
    if (assertion.failureMessages && Array.isArray(assertion.failureMessages) && assertion.failureMessages.length > 0) {
      return assertion.failureMessages[0];
    }
    return null;
  }

  /**
   * Извлекает стек вызовов из результата Jest
   * @param assertion Результат теста от Jest
   * @returns Стек вызовов или null
   */
  private getStackTraceFromJest(assertion: any): string | null {
    if (assertion.failureMessages && Array.isArray(assertion.failureMessages) && assertion.failureMessages.length > 1) {
      return assertion.failureMessages.slice(1).join('\n');
    }
    return null;
  }
}
